### ISSUE TITLE
deprecate `expr_free_symbols`

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#21494]
This is a general property that was added to solve a very
specific problem but it adds a layer of abstraction that is
not necessary in general.

1. objects that have structural "non-expression" nodes already
   allow one to focus on on the expression node if desired, e.g.
```python
   >>> Derivative(f(x), x).expr
   f(x)
```
   introduction of this property encourages imprecise thinking
   when requesting free_symbols since it allows one to get
   symbols from a specific node of an object without specifying
   the node

2. it incorrectly added the property to AtomicExpr so numbers
   are returned as expr_free_symbols
```python
   >>> S(2).expr_free_symbols
   2
```
3. the application of the concept was misapplied to define
   Subs.expr_free_symbols: it added in expr_free_symbols of
   the point but the point is a Tuple so nothing was added

4. it is not used anywhere else in the code base except in the
   context of differentiating a Subs object which suggests that
   it is not something of general use, this is also confirmed by
   the fact that,

5. it was added without specific tests except for test of the
   derivatives of the Subs object for which it was introduced

**Comment by Upabjojr** [comment]
`expr_free_symbols` is meant to represent indexed objects such as `MatrixElement` and `Indexed` as free symbols. Currently `.free_symbols` keeps iterating over their indices.

If I have `A[i, j]` in an expression, I would like to have a way to return `{A[i, j]}` as a free symbol, and not `{A, i, j}`, which by itself doesn't make sense.

Unfortunately much of the code has already been written in order to use `{A, i, j}` as free symbols, so it's easier to introduce a new method rather than fixing the old one.

I this `.expr_free_symbols` can be useful.

**Comment by oscarbenjamin** [comment]
This deprecation has not been released yet. There needs to be a clear decision before the imminent release.

The points raised by @smichr are good ones but an alternative to deprecation would be fixing the bugs and completing the implementation.

Generally though I would prefer not to keep adding attributes and methods to Expr which already has far too many. This would be better as a public function even if it uses an internal method to allow extensibility by different classes. At least the implementation should be complete before adding any new API to Expr/Basic because it creates future problems with the need for incompatible changes.

I suppose the need for something like `expr_free_symbols` is related to #20809 and #20809.

**Comment by smichr** [comment]
> There needs to be a clear decision before the imminent release.

Please get rid of the current implementation for all the reasons I cite. And perhaps create a new method `free_expressions` (after thorough discussion) that does not conflate symbols and non-symbols.

**Comment by oscarbenjamin** [comment]
I'll take a look at this later.

I suppose that there needs to be a better fix in Piecewise rather than requiring monkeypatching. There are lots of things wrong with `Piecewise.eval`. It calls `simplify` even under `evaluate(False)`. This close to the release is not a great time to make big changes though.

**Comment by bocklund** [comment]
Thanks for looking. In our code, the only change from the SymPy 1.2 version was to change the access of `c.free_symbols` to `c.expr_free_symbols`. 

https://github.com/sympy/sympy/blob/9dd704b071e5134a0e8beeeb1a2d2599536f5144/sympy/functions/elementary/piecewise.py#L175

If we could get the same behavior of `expr_free_symbols` in this part of `Piecewise.eval` instead of `free_symbols`, that should solve it for us. We could remove our monkey patch and usage of `expr_free_symbols` and our users could retain the performance that we have with `expr_free_symbols`.

**Comment by oscarbenjamin** [comment]
> Locally our test suite goes from about 2 minutes on my machine to about 5 minutes. With some profiling, I was able to identify the time spent in our test suite in `subs` grew from 7% to 20% of our test suite time and `xreplace` grew from to 11% to 40% of our test suite time.

Do you think that you could extract a relatively simple reproducer of the slowness? I think it would be worth investigating. I guess subs is causing Piecewise.eval to rerun but it would be good to unpick what exactly is slow in there.

**Comment by Upabjojr** [comment]
> Turning off the monkey patch or using `free_symbols` instead of `expr_free_symbols` is also a significant performance regression for us, I think related to the large Piecewise expressions generated in pycalphad.

Since Python 3.4 there is support for [single dispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) which can be used as a better alternative to monkey patching.

**Comment by asmeurer** [comment]
What was the decision here? I remember expr_free_symbols having a specific use case for differentiating non-Symbol objects. I think we should at least remove the deprecation before this is discussed further. If we do keep it, let's make sure it is documented properly, as clearly this was not done well enough previously. 

Regarding @bocklund's use-case we should figure out how to make that work without monkeypatching. Oscar is going to be making a big push for performance improvements in the very near future, please open any issues for performance related things like this (or in any cases where you find yourself forced to monkeypatch SymPy) so that we can get them fixed right.

**Comment by smichr** [comment]
> I remember expr_free_symbols having a specific use case for differentiating non-Symbol objects.

I took care of that in #21485. I listed in the OP all the reasons to deprecate this method.

**Comment by oscarbenjamin** [comment]
Perhaps it would be good just to show how to implement `expr_free_symbols` as a function for anyone previously using it.

**Comment by Upabjojr** [comment]
> Perhaps it would be good just to show how to implement `expr_free_symbols` as a function for anyone previously using it.

What about putting `expr_free_symbols` into `sympy.sandbox`? The sandbox module was created as a place to put _work in progress_ code or code that is not sound but still useful.

**Comment by asmeurer** [comment]
@Upabjojr does expr_free_symbols add anything that isn't already provided by _diff_wrt? It seems to me that any logic that would use expr_free_symbols should just walk the expression tree and check for _diff_wrt expressions (which could be abstracted by a helper function). 

It seems that either way, the method was poorly named. It is only needed for differentiation, but is named like the more general free_symbols, which applies to more than just derivatives (and in fact, should really not be directly tied to derivatives because of generalized differentiation wrt variable expressions). 

If it is needed, I don't see why it should be a function rather than an attribute. The whole point of it is that specific objects can define themselves as "free symbols". That means it needs to be an attribute on the class.

**Comment by oscarbenjamin** [comment]
> If it is needed, I don't see why it should be a function rather than an attribute. The whole point of it is that specific objects can define themselves as "free symbols". That means it needs to be an attribute on the class.

There are many attributes/methods in SymPy that should really just be functions e.g.:
```python
def as_coeff_Add(expr):
    if isinstance(expr, Add) and isinstance(expr.args[0], Number):
        return expr.args[0], expr._new_rawargs(*expr.args[1:])
    elif isinstance(expr, Number):
        return expr, S.Zero
    else:
        return S.Zero, expr
```
The reasons I suggest writing a function equivalent for `expr_free_symbols` are

1. The function can be used in downstream libraries.
2. It makes it clear what the function does and that actually it can probably be replaced trivially by something simpler anyway.

The actual problem here is #14933 which can definitely be fixed in a better way but it might be too late to fix properly in 1.9. In that case a quick workaround replacement for `expr_free_symbols` would be useful. In general it is good to provide an alternative in a deprecation issue.

**Comment by smichr** [comment]
> It is only needed for differentiation

It is no longer needed for differentiation. It is no longer used in the codebase. It only appears as a method to be removed. I will take a look at the Piecewise issue to see why it works there when replacing `free_symbols`.

**Comment by asmeurer** [comment]
My worry is if there was more logic that just wasn't implemented yet. If it is just Subs, then I agree we don't need it, but if it's also needed for indexed derivatives that aren't implemented yet, then we should consider if we aren't going backwards by preventing that (which is why I wanted @Upabjojr to comment on this either way).

**Comment by smichr** [comment]
> but if it's also needed for indexed derivatives that aren't implemented yet

OK, but also consider that this might be a kludge instead of defining a more precise extraction from a node (see OP comment 1).

**Comment by smichr** [comment]
@oscarbenjamin , removal can be delayed for another release as far as I'm concerned.

**Comment by asmeurer** [comment]
Another thing to consider is if the only point of this API is just to short-circuit expressions whose derivative should be 0. I don't know if a whole API is needed for just that optimization.

**Comment by asmeurer** [comment]
I think we should definitely delay the removal. That gives us time to figure out the Piecewise issue right, and also to figure out if we really want to deprecate it or not. And it's not like some other feature is blocked on its removal, so there's no downside to postponing the deprecation.

**Comment by oscarbenjamin** [comment]
>  removal can be delayed for another release as far as I'm concerned.

Okay, let's do that. I'll open a PR for 1.9 to revert the deprecation warnings. We could just leave it undeprecated on master if we think that this can be resolved by 1.10 but there would need to be a release blocker issue for 1.10.

**Comment by oscarbenjamin** [comment]
I've opened #22158 which undeprecates `expr_free_symbols` in the 1.9 branch.

**Comment by Upabjojr** [comment]
> My worry is if there was more logic that just wasn't implemented yet. If it is just Subs, then I agree we don't need it, but if it's also needed for indexed derivatives that aren't implemented yet, then we should consider if we aren't going backwards by preventing that (which is why I wanted @Upabjojr to comment on this either way).

It would be nice to make the equation solver work with indexed symbols. It does currently not work because `.free_symbols` is used in various locations to test the presence of variables. An equivalent method returning the indexed object without its subcomponents is probably needed.

**Comment by Upabjojr** [comment]
Mmm... actually I've just noticed that using the solver with both `Indexed` and `MatrixElement` works in the latest release. I think the original problem for which I introduced `.expr_free_symbols` has been fixed.

I therefore change my stand in favor of deprecation.
